packages/core/src/external/mpi/match-patients.ts

```diff
@@ -0,0 +1,147 @@
+import { intersectionWith, isEqual } from "lodash";
+import { PatientDataMPI } from "./patient";
+import { normalizePatientDataMPI } from "./normalize-patient";
+import jaroWinkler from "jaro-winkler";
+
+// Define a type for the similarity function
+type SimilarityFunction = (
+  patient1: PatientDataMPI,
+  patient2: PatientDataMPI,
+  threshold: number
+) => boolean;
+
+type MatchingRule = (patient1: PatientDataMPI, patient2: PatientDataMPI) => boolean;
+
+/**
+ * `matchPatients` filters patients based on a similarity function, rules, and a threshold.
+ * @param similarityFunction - Determines if a patient is a match.
+ * @param matchingRules - Rules that determines if a patient is a match.
+ * @param patients - Array of patients.
+ * @param demo - Patient data to find matches for.
+ * @param threshold - Minimum similarity score for a match.
+ * @param greedy - If true, returns the first match. If false, returns all matches.
+ * @returns matched patients.
+ */
+export const matchPatients = (
+  similarityFunction: SimilarityFunction,
+  matchingRules: MatchingRule[],
+  patients: PatientDataMPI[],
+  demo: PatientDataMPI,
+  threshold: number,
+  greedy = true
+): PatientDataMPI[] => {
+  const matchFunction = (patient: PatientDataMPI) => {
+    const PatientDataMPI = normalizePatientDataMPI(patient);
+    if (!PatientDataMPI) {
+      return false;
+    }
+    for (const rule of matchingRules) {
+      if (rule(demo, PatientDataMPI)) {
+        return true;
+      }
+    }
+    return similarityFunction(PatientDataMPI, demo, threshold);
+  };
+  if (greedy) {
+    const foundPatient = patients.find(matchFunction);
+    return foundPatient ? [foundPatient] : [];
+  } else {
+    return patients.filter(matchFunction);
+  }
+};
+
+/**
+ * This function checks if the patient has any personal identifiers that match the demo. The idea of having rules
+ * is that rules will override statistical similarity methods. This rule should act as an override.
+ * @param demo
+ * @param patient
+ * @returns true if the patient has any personal identifiers that match the demo.
+ */
+export const matchingPersonalIdentifiersRule = (
+  demo: PatientDataMPI,
+  patient: PatientDataMPI
+): boolean => {
+  return Boolean(
+    intersectionWith(patient.personalIdentifiers || [], demo.personalIdentifiers || [], isEqual)
+      .length > 0
+  );
+};
+
+/**
+ * This function checks if the patient has any contact details (phone or email) that match the demo. The idea of this rule
+ * us that contact info is a very strong unique identifier if they do match, but doesn't necessarily mean that the patients
+ * don't match if they don't match.
+ * @param demo
+ * @param patient
+ * @returns true if the patient has any contact details that match the demo.
+ */
+export const matchingContactDetailsRule = (
+  demo: PatientDataMPI,
+  patient: PatientDataMPI
```

Comment:
I suggest we name them as patient1 and patient2... calling demo and patient makes the reader spend cycles wondering what's the difference between them